[16.0][MIG] mail_restrict_follower_selection: Migration to 16.0

AungKoKoLin1997 opened the pull request:
@qrtl
In this migration , there is a issue in adding domain for `partner_ids` in adding new followers. We need to refresh for each document model.
Related issue: https://github.com/odoo/odoo/issues/109601

hbrunn committed:
[ADD] mail_restrict_follower_selection

hbrunn committed:
[FIX] better default filter

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

bouvyd committed:
[MIG] mail_restrict_follower_selection: Migrated to 10.0

phucngta committed:
[11.0][MIG] mail_restrict_follower_selection

etobella committed:
[IMP] disable the option to create followers that do not meet the conditions

didierdonze committed:
Translated using Weblate (French) [ci skip]

Currently translated at 100.0% (3 of 3 strings)

Translation: social-11.0/social-11.0-mail_restrict_follower_selection
Translate-URL: https://translation.odoo-community.org/projects/social-11-0/social-11-0-mail_restrict_follower_selection/fr/

marylla committed:
Translated using Weblate (German)

Currently translated at 100.0% (3 of 3 strings)

Translation: social-11.0/social-11.0-mail_restrict_follower_selection
Translate-URL: https://translation.odoo-community.org/projects/social-11-0/social-11-0-mail_restrict_follower_selection/de/

enriquemartin committed:
[FIX] mail_restrict_follower_selection: res_model not always in context
When creating a record from a record from another model, the model is not in the context (`default_res_model` key). For example: creating an invoice from a sale order.

JordiBForgeFlow committed:
fixes a corner case where you could send an email to a partner
that you are not supposed to.

etobella committed:
[12.0][MIG] mail_restrict_follower_selection

JoanSForgeFlow committed:
[IMP] mail_restrict_follower_selection: black, isort

JoanSForgeFlow committed:
[MIG] mail_restrict_follower_selection : Migration to 13.0

Tardo committed:
[IMP] mail_restrict_follower_selection: Add condition for tests

Tardo committed:
[IMP] mail_restrict_follower_selection: Add condition for tests

alessandro-fiorino committed:
Translated using Weblate (Italian)

Currently translated at 75.0% (3 of 4 strings)

Translation: social-13.0/social-13.0-mail_restrict_follower_selection
Translate-URL: https://translation.odoo-community.org/projects/social-13-0/social-13-0-mail_restrict_follower_selection/it/

angelmoya committed:
[IMP] mail_restrict_follower_selection: Adding tests and improving pre-commit

Commit:
[IMP] pre-commit run -a

Tardo committed:
[IMP] Apply pre-commit changes: Resolve conflicts

olgamarcocb committed:
[MIG] mail_restrict_follower_selection: Migration to 14.0

oca-travis committed:
[UPD] Update mail_restrict_follower_selection.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: social-14.0/social-14.0-mail_restrict_follower_selection
Translate-URL: https://translation.odoo-community.org/projects/social-14-0/social-14-0-mail_restrict_follower_selection/

AungKoKoLin1997 committed:
[IMP] mail_restrict_follower_selection: black, isort, prettier

AungKoKoLin1997 committed:
[15.0][MIG] mail_restrict_follower_selection: Migration to 15.0

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (7 of 7 strings)

Translation: social-14.0/social-14.0-mail_restrict_follower_selection
Translate-URL: https://translation.odoo-community.org/projects/social-14-0/social-14-0-mail_restrict_follower_selection/it/

OCA-git-bot committed:
[IMP] update dotfiles [ci skip]

AungKoKoLin1997 committed:
[MIG] mail_restrict_follower_selection: Migration to 16.0

yostashiro commented on mail_restrict_follower_selection/models/mail_wizard_invite.py:
```diff
@@ -0,0 +1,43 @@
+        for field in arch.xpath('//field[@name="partner_ids"]'):
+            domain = self._mail_restrict_follower_selection_get_domain()
+            domain = ast.literal_eval(domain)
+            domain.append(("type", "!=", "private"))
```
Can we get this domain from `self._fields["partner_ids"].domain` (and use expression.AND(), perhaps)?

I think this domain manipulation should be done inside _mail_restrict_follower_selection_get_domain() instead of here.

yostashiro commented on mail_restrict_follower_selection/models/mail_wizard_invite.py:
```diff
@@ -0,0 +1,47 @@
+        )
+        arch = etree.fromstring(result["arch"])
+        for field in arch.xpath('//field[@name="partner_ids"]'):
+            field.attrib["domain"] = str(
```
@AungKoKoLin1997 Do we need this str conversion?

AungKoKoLin1997 replied:
Yes. We need str conversion because we can't assign list to domain attribute.

hieulucky111 commented on mail_restrict_follower_selection/models/mail_thread.py:
```diff
@@ -0,0 +1,30 @@
+        for key in result:
+            for partner_id, email, lang, reason in result[key]:
+                if partner_id:
+                    partner = self.env["res.partner"].search(
```
should be `search_count` to improve performance

hieulucky111 reviewed (approved):
LGTM.

hieulucky111 reviewed (approved):
Code review
LGTM

yostashiro commented:
https://github.com/OCA/social/issues/1070

@AungKoKoLin1997 Let's see how people respond to this, and make adjustments in this PR if there are no strong concerns.

AungKoKoLin1997 committed:
[IMP] adjustments
1. Use forecreate=false in mail_restrict_follower_selection/data/ir_config_parameter.xml so that a deleted configuration is not recreated by a module update
2. Change the default domain to something less agressive
3. Add an example restriction restricting following a specific model

michaelslade54 commented on mail_restrict_follower_selection/models/mail_wizard_invite.py:
```diff
@@ -0,0 +1,47 @@
+        return domain
+
+    @api.model
+    def get_view(self, view_id=None, view_type="form", **options):
```
This get_view method is not called in every instance, so the domain for the partner_ids field is not set correctly.

get_view is only called in the first instance, after that, onchange method is called, even after changing models.

If you add a follower in one model, then go to another model with a different domain set and add a follower again, the get_view function is not called, leading to an incorrect domain set on partner_ids.

Hope this is clear.

michaelslade54 commented:
Not withstanding the issue with the domain for `partner_ids` not reloading when changing models, while a user **not** within the required domain is not added as a follower as intended, they are still sent an email informing them that they have an invitation to follow, which is undesirable.

Can someone please confirm?

I believe the domain needs to be added to `new_partners` here...
https://github.com/odoo/odoo/blob/a6f8d9c512ffb816e18bb43a5efa510707877cf6/addons/mail/wizard/mail_wizard_invite.py#L56

AungKoKoLin1997 commented:
> I believe the domain needs to be added to new_partners here...
> https://github.com/odoo/odoo/blob/a6f8d9c512ffb816e18bb43a5efa510707877cf6/addons/mail/wizard/mail_wizard_invite.py#L56

@michaelslade54 If I am not mistaken, `add_followers` is called after we choose partners in wizard and click button. What I mentioned the issue was when adding domain to partner_ids.

fkantelberg reviewed (approved):
Tested and shortly reviews. LGTM

I think that the missing domain refresh on model swap is outside of the scope of this PR.

NL66278 reviewed (approved):
:+1: LGTM Could have removed copyright from init files, but not blocking.

hbrunn commented:
/ocabot merge nobump
